src/cursor/cur_layered.c

```diff
@@ -206,42 +206,229 @@ __clayered_close_cursors(WT_CURSOR_LAYERED *clayered, bool include_ingest)
+    }
+
+    return (ret);
+}
+
 /*
  * __clayered_adjust_state --
  *     Update the state of the cursor to match the state of the disaggregated system. In particular,
- *     if the system has changed in a way that makes constituent cursors out of date, close them.
- *     They will be opened later as needed.
+ *     if the system has changed in a way that makes constituent cursors out of date, either reopen
+ *     them or close them, and let them be opened later as needed.
  */
 static int
-__clayered_adjust_state(WT_CURSOR_LAYERED *clayered, bool *state_updated)
+__clayered_adjust_state(WT_CURSOR_LAYERED *clayered, bool iteration, bool *state_updated)
 {
     WT_CONNECTION_IMPL *conn;
-    uint64_t current_checkpoint_id;
-    bool current_leader;
+    WT_CURSOR *old_stable;
+    WT_SESSION_IMPL *session;
+    WT_TXN_SHARED *txn_shared;
+    uint64_t current_checkpoint_id, snapshot_gen;
+    bool change_ingest, change_stable, current_leader;
 
-    conn = S2C(CUR2S(clayered));
-    current_leader = conn->layered_table_manager.leader;
     *state_updated = false;
-
+    session = CUR2S(clayered);
+    conn = S2C(session);
+    current_leader = conn->layered_table_manager.leader;
     /* Get the current checkpoint id. This only matters if we are a follower. */
     if (!current_leader)
         WT_ACQUIRE_READ(current_checkpoint_id, conn->disaggregated_storage.global_checkpoint_id);
     else
-        current_checkpoint_id = 0;
+        current_checkpoint_id = WT_DISAGG_CHECKPOINT_ID_NONE;
+
+    if (current_leader != clayered->leader || current_checkpoint_id != clayered->checkpoint_id) {
+        change_ingest = change_stable = false;
+        snapshot_gen = clayered->snapshot_gen;
+
+        if (current_leader != clayered->leader) {
+            /*
+             * If leader state has changed, all cursors are affected.
+             *
+             * FIXME-SLS-1449 Maybe this can be relaxed - a stable cursor might be left open on
+             * downgrade as long as it does readonly operations from then on.
+             */
+            change_ingest = true;
+            change_stable = true;
+        } else {
+            /*
+             * We have a new checkpoint on the follower. We'd like to reopen the stable cursor, but
+             * we must abide by cursor and transactional semantics.
+             *
+             * First off, if this is an iteration, we won't upgrade the cursor.
+             */
+            if (iteration)
+                return (0);
 
-    /* If leader state has changed, close all the cursors. */
-    if (current_leader != clayered->leader) {
-        WT_RET(__clayered_close_cursors(clayered, true));
+            /*
+             * Layered cursors are sometimes paired with read timestamps. When using read
+             * timestamps, it's always safe to update cursors, even during iterations. That's
+             * because the view at a timestamp is always consistent, the history store covers that.
```

Comment:
So why we disable update for iteration even with read timestamp?

Reply:
Ah thanks!  It should always allow updates with read timestamps.  I'll fix.